Aplib.Core/ExposedQueue.cs

```
@@ -0,0 +1,193 @@
+    {
+        /// <summary>
+        /// The length of the array.
+        /// </summary>
+        public int MaxCount { get; private set; }
+
+        /// <summary>
+        /// Actual number of elements in the array.
+        /// Limited to <see cref="MaxCount"/>.
+        /// </summary>
+        public int Count
+        {
+            get => _count;
+            private set => _count = value > MaxCount ? MaxCount : value;
+        }
+
+        /// <inheritdoc/>
+        public bool IsReadOnly => false;
+
+        private int _count;
+
+        private readonly T[] _array;
+        private int _head;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExposedQueue{T}"/> class.
+        /// </summary>
+        /// <param name="size">The size of the array.</param>
+        public ExposedQueue(int size)
+        {
+            MaxCount = size;
+            _array = new T[MaxCount];
+            _head = MaxCount - 1;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExposedQueue{T}"/> class.
+        /// By default, assumes the array is filled.
+        /// </summary>
+        /// <param name="array">An array to use as the circular array.</param>
+        /// <param name="count">The number of actual elements in the array.</param>
+        public ExposedQueue(T[] array, int count = -1)
+        {
+            MaxCount = array.Length;
+            _array = array;
+            _head = MaxCount - 1;
+            Count = count == -1 ? MaxCount : count;
+        }
+
+        /// <summary>
+        /// Gets the element at the specified index.
+        /// </summary>
+        /// <param name="index">The index of the element to get.</param>
+        /// <returns>The element at the specified index.</returns>
+        public T this[int index]
+        {
+            get => _array[(index + _head + 1) % MaxCount];
+            private set => _array[(index + _head + 1) % MaxCount] = value;
+        }
```

Could you throw an error when the specified index is higher than the Count? I don't think we want to make the index 'wrappable' like that.